onnxruntime/core/xnnpack/max_pooling_2d.cc

```diff
@@ -0,0 +1,109 @@
+  ORT_ENFORCE(info.GetAttr("input_padding_right", &input_padding_right).IsOK());
+  ORT_ENFORCE(info.GetAttr("input_padding_bottom", &input_padding_bottom).IsOK());
+  ORT_ENFORCE(info.GetAttr("input_padding_left", &input_padding_left).IsOK());
+  ORT_ENFORCE(info.GetAttr("pooling_height", &pooling_height).IsOK());
+  ORT_ENFORCE(info.GetAttr("pooling_width", &pooling_width).IsOK());
+  ORT_ENFORCE(info.GetAttr("stride_height", &stride_height).IsOK());
+  ORT_ENFORCE(info.GetAttr("stride_width", &stride_width).IsOK());
+  ORT_ENFORCE(info.GetAttr("dilation_height", &dilation_height).IsOK());
+  ORT_ENFORCE(info.GetAttr("dilation_width", &dilation_width).IsOK());
+
+  ORT_ENFORCE(info.GetAttr("output_min", &output_min).IsOK());
+  ORT_ENFORCE(info.GetAttr("output_max", &output_max).IsOK());
+  ORT_ENFORCE(info.GetAttr("padding_mode", &padding_mode).IsOK());
+
+  input_padding_top_ = gsl::narrow<uint32_t>(input_padding_top);
+  input_padding_right_ = gsl::narrow<uint32_t>(input_padding_right);
+  input_padding_bottom_ = gsl::narrow<uint32_t>(input_padding_bottom);
+  input_padding_left_ = gsl::narrow<uint32_t>(input_padding_left);
+  pooling_height_ = gsl::narrow<uint32_t>(pooling_height);
+  pooling_width_ = gsl::narrow<uint32_t>(pooling_width);
+  stride_height_ = gsl::narrow<uint32_t>(stride_height);
+  stride_width_ = gsl::narrow<uint32_t>(stride_width);
+  dilation_height_ = gsl::narrow<uint32_t>(dilation_height);
+  dilation_width_ = gsl::narrow<uint32_t>(dilation_width);
+  uint32_t flags = 0;
+  if (padding_mode == 1) {
+    flags |= XNN_FLAG_TENSORFLOW_SAME_PADDING;
+  }
+  int64_t input_channels = input_type_proto->tensor_type().shape().dim(3).dim_value();
+
+  xnn_status status;
+  struct xnn_operator* p;
+  status = xnn_create_max_pooling2d_nhwc_f32(input_padding_top_, input_padding_right_, input_padding_bottom_,
+                                             input_padding_left_, pooling_height_, pooling_width_, stride_height_,
+                                             stride_width_, dilation_height_, dilation_width_,
+                                             static_cast<size_t>(input_channels), static_cast<size_t>(input_channels),
+                                             static_cast<size_t>(input_channels), output_min, output_max, flags, &p);
+  ORT_ENFORCE(status == xnn_status_success);
+  op0_.reset(p);
+}
+
+Status MaxPool2D::Compute(OpKernelContext* context) const {
+  const auto* X = context->Input<Tensor>(0);
+  Tensor* Y = nullptr;
+  if (has_const_output_shape_) {
+    Y = context->Output(0, output_shape_);
+  } else {
+    return Status(common::ONNXRUNTIME, common::NOT_IMPLEMENTED);
+  }
+
+  const TensorShape& input_shape = X->Shape();
+  xnn_status status = xnn_setup_max_pooling2d_nhwc_f32(
+      op0_.get(), input_shape[0] /* batch size */, input_shape[1] /* input height */, input_shape[2] /* input width */,
+      X->Data<float>() /* input */, Y->MutableData<float>() /* output */, nullptr /* threadpool */);
```

Comment:
>threadpool [](http://example.com/codeflow?start=85&length=10)

It will be single-threaded under the hood?
Another question would be that if we do created a pthreadpool here, it must be static?

Reply:
I will need to get a pthreadpool  pointer from some there.  I don't know how to do it yet.